src/python/pants/jvm/target_types.py

```diff
@@ -330,9 +372,9 @@ class JunitTestExtraEnvVarsField(TestExtraEnvVarsField):
 # -----------------------------------------------------------------------------------------------
 
 
-class JvmMainClassNameField(StringField):
-    alias = "main"
+class JvmRequiredMainClassNameField(JvmMainClassNameField):
```

Comment:
I've made this a subclass rather than a standalone class to reduce churn elsewhere in the codebase.

My instinct is that I could make the new `_find_main_by_javap` helper into a fully-fledged rule, and we could use that when constructing `deploy_jar`s if `main` is not specified. When that happens, this subclass can just go away.